### PR TITLE
Switch Jenkins.getInstance() to @Nonnull

### PR DESCRIPTION
**Pull request by batmat**
As [detailed](https://github.com/jenkinsci/jenkins/commit/96150d3a465f428b5fbc5d03c8cd00115f4749c9#commitcomment-17288107) by @stephenc:

> So there is `getInstanceOrNull()` which is for use in Jenkins core or
> in code that plugins add that may escape the Jenkins singleton
> lifecycle.
> 
> In regular plugins `getInstance()` will never return null as the
> lifecycle does not instantiate them until after there is a Jenkins
> singleton and the plugins will be stopped before there is no
> singleton.
> 
> The `getActiveInstance()` is therefore equivalent to `getInstance()`
> and should never have been born.
> 
> The only way a plugin can get a `null` from `Jenkins.getInstance()` is
> if you install an `atExit()` handler, use a `PhantomReference` type
> thing, or directly manipulate the servlet container in some way or
> other. A correctly written plugin should unwind any of those things
> when it is stopped or a method bound to the termination lifecycle. If
> there are some cases (for which I cannot anticipate a good coder being
> able to write a correct termination lifecycle method to fix) where the
> code cannot be unhooked... then in those cases the plugin should use
> `getInstanceOrNull()` and know how to respond to the `null` that could
> not be avoided.

### CODE REVIEW

**Comment by batmat** [comment]
Ping @jenkinsci/code-reviewers

**Comment by oleg-nenashev** [comment]
I'm not sure I follow the idea. Do we want to break the compatibility in 2.x? The method may be used in the code outside plugins (e.g. Boot hook scripts), hence I'm still not convinced it's safe.

**Comment by batmat** [comment]
Not absolutely sure what you have in mind with "boot hook scripts".
If you talk about init.d groovy scripts to be called during startup, then I created https://github.com/batmat/jenkins-getinstance-tester to check. 

This shows that even in that case `Jenkins.getInstance()` is not null. (well, as it's used generally to configure plugins, that's a good news. So possibly, you're talking of something else, now I think of it again).

```
Apr 29, 2016 1:26:27 PM org.jenkinsci.main.modules.sshd.SSHD start
INFO: Started SSHD at port 42168
Well, OK, getInstance() did not return null.
Apr 29, 2016 1:26:27 PM jenkins.util.groovy.GroovyHookScript execute
```

**Comment by oleg-nenashev** [comment]
I also need to dig into the code

**Comment by daniel-beck** [comment]
Tentative :+1: here, but more @jenkinsci/code-reviewers would be welcome.

**Comment by stephenc** [comment]
I'm 👍 though it might be better to have a system property control the exact behaviour for 10-20 versions so that it's annotatied `@NonNull` but it throwing the ISE can be turned off by a system property to allow users an escape hatch if they have a valid code path (or even an invalid one) that needs addressing somehow

**Comment by stephenc** [comment]
(to be clear I would have the default throw the ISE)

**Comment by daniel-beck** [comment]
Right, we're approaching 2.x LTS now, so that would be a useful addition.

**Comment by batmat** [comment]
Right. Good idea. Will be able to do that on Sunday evening.

**Comment by batmat** [comment]
@stephenc Just amended the version using a sysprop as you asked rightly. 
And also, yes, it's throwing an exception by default as it's the very goal of that change indeed.

Ping @daniel-beck @oleg-nenashev also for a second look, and @jenkinsci/code-reviewers in general since this (small) change needs eyes.

**Comment by stephenc** [comment]
👍

**Comment by daniel-beck** [comment]
:+1:

**Comment by batmat** [comment]
Will merge later today or tomorrow if nobody objects in the meantime.

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 1]
```diff
@@ -753,10 +753,17 @@ public static Jenkins getInstanceOrNull() {
...
+    @Nonnull
     public static Jenkins getInstance() {
-        // TODO throw an IllegalStateException in Jenkins 2.0+
-        return HOLDER.getInstance();
+        Jenkins instance = HOLDER.getInstance();
+        if (instance == null) {
+            if(!Boolean.getBoolean(Jenkins.class.getName()+".disableExceptionOnNullInstance")) {
+                // remove that second block around 2.20 (that is: ~20 versions to battle test it)
+                // See https://github.com/jenkinsci/jenkins/pull/2297#issuecomment-216710150
+                throw new IllegalStateException();
```
Comment (oleg-nenashev) [inline core/src/main/java/jenkins/model/Jenkins.java]: Would be useful to restore the original text then

**Comment by oleg-nenashev** [comment]
IMHO needs an exception text to be restored, hence weak 👎 . Minor fix BTW

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 1 above)
Comment (batmat) [inline core/src/main/java/jenkins/model/Jenkins.java]: Can you elaborate @oleg-nenashev ? Not sure which original text you're talking about. 

`// TODO throw an IllegalStateException in Jenkins 2.0+` ?

Comment (stephenc) [inline core/src/main/java/jenkins/model/Jenkins.java]: I think he wants

```
throw new IllegalStateException("Jenkins has not been started, or was already shut down");
```

File: core/src/main/java/jenkins/model/Jenkins.java [hunk 2]
```diff
@@ -753,10 +753,17 @@ public static Jenkins getInstanceOrNull() {
...
     @CLIResolver
-    @Nullable // TODO replace with non-null once Jenkins 2.0+
+    @Nonnull
     public static Jenkins getInstance() {
-        // TODO throw an IllegalStateException in Jenkins 2.0+
-        return HOLDER.getInstance();
+        Jenkins instance = HOLDER.getInstance();
+        if (instance == null) {
+            if(!Boolean.getBoolean(Jenkins.class.getName()+".disableExceptionOnNullInstance")) {
+                // remove that second block around 2.20 (that is: ~20 versions to battle test it)
```
Comment (stephenc) [inline core/src/main/java/jenkins/model/Jenkins.java]: This should have `TODO` so we find it again

Comment (batmat) [inline core/src/main/java/jenkins/model/Jenkins.java]: Done.

File: core/src/main/java/jenkins/model/Jenkins.java (hunk 1 above)
Comment (batmat) [inline core/src/main/java/jenkins/model/Jenkins.java]: Right, like in current `getActiveInstance()`.
Fixed.

**Comment by oleg-nenashev** [comment]
👍

**Comment by batmat** [comment]
@oleg-nenashev Fixed, thanks for the review.

**Comment by batmat** [comment]
Lol, commenting at the very same moment. o/ @oleg-nenashev ;-)

**Comment by batmat** [comment]
Will merge after the build completes successfully. 

Thanks to everyone.

**Comment by oleg-nenashev** [comment]
@batmat A kind of lifehack: Merge&Squash also allows to avoid a adding a merge commit for single-commit PR. Makes the history shorter: https://github.com/jenkinsci/jenkins/commits/master

/me hides before he starts getting grumbles from non-squash fans

**Comment by daniel-beck** [comment]
@oleg-nenashev *You* fix the changelog after every release then. Too much crap to sift through otherwise, having commits titled `Merge pull request` speeds that up considerably for me.

**Comment by oleg-nenashev** [comment]
@daniel-beck Yes, that's a valid point. May also help if we start squashing commits, but oh well

**Comment by oleg-nenashev** [comment]
@daniel-beck But I've got the point that I should not forget about changelogs for PRs I merge. My impression was that it's better to process it before the release because somebody has to do it in any case. :(

**Comment by daniel-beck** [comment]
@oleg-nenashev Yep, and that someone has dozens of commits to sift through. If every PR has a merge commit with the generated message, that makes it much easier.

**Comment by batmat** [comment]
Gonna modify the changelog.

**Comment by batmat** [comment]
@daniel-beck is that how it must be done https://github.com/batmat/jenkins/commit/2b6fb33365bb420017c4debd23f7a788388347f8?
(Mostly the format, I will re-check the content later today, was a bit in a hurry just now).

(It's on my fork to check first)

Then IIRC, I just push that onto the jenkins one, w/o PR right?

Thanks

**Comment by batmat** [comment]
:fire: [JENKINS-34857](https://issues.jenkins-ci.org/browse/JENKINS-34857)

TL;DR: someone started having issue with this PR on Jenkins 2.4 released yesterday... So possibly a urgent revert + release may be needed. Mainly because (still supposedly at this stage of analysis) the code is running in an agent hence the sysprop isn't present there :-/.

The associated crashing code for reference: https://github.com/jenkinsci/subversion-plugin/blob/master/src/main/java/hudson/scm/SubversionSCM.java#L2544 (cc @recena)

**Comment by stephenc** [comment]
That is indicative of a real issue in the subversion plugin. Plugins should not be pulling the `Jenkins` class over remoting. IIRC we had plans to block that at some point in the future.

**Comment by recena** [comment]
`<ironic mode=on>`Without any doubt this is the best way to find out a bug.`</ironic>`

**Comment by batmat** [comment]
Possibly @recena the short term "fix" could be to call `Jenkins.getInstanceOrNull()`. But the issue is it's only starting at 1.657, so you may want to call this method by reflection, and then fall back onto `Jenkins.getInstance()` if not found. Not beautiful, but should prevent the SVN plugin to crash w/ 2.4 (or if we reintroduce the exception in the future).

**Comment by stephenc** [comment]
Well in actual fact that method could be better served as

```
@CheckForNull
    private static DescriptorImpl descriptor() {
        return ExtensionList.lookup(Descriptor.class).get(DescriptorImpl.class);
    }
```

But more critically, you should not be looking up the descriptor in a remote callable

**Comment by stephenc** [comment]
Plus https://github.com/jenkinsci/subversion-plugin/blob/master/src/main/java/hudson/scm/SubversionSCM.java#L1041 is going to be pulling too much code over to the slave. Instead the options should be created https://github.com/jenkinsci/subversion-plugin/blob/master/src/main/java/hudson/scm/SubversionSCM.java#L948 and modified for the fact that they are going to be for the remote agent...

Basically the Subversion plugin is doing this all wrong and causing unnecessary classloading into the slaves which may be having other impacts that have remained hidden from us
